src/app.js

```diff
@@ -211,40 +213,107 @@ export default function (app) {
       var realValue = element[name]
 
       if (name === "onUpdate") {
-        defer(value, element)
+          defer(value, element)
 
       } else if (value !== oldValue || realValue !== value) {
         setElementData(element, name, value, oldValue)
       }
     }
   }
 
-  function patch(parent, element, oldNode, node) {
+  function batchRemove (parent, element, node) {
+      batch.push(parent.removeChild.bind(parent, element))
+      if (node.data && node.data.onRemove) {
+        defer(node.data.onRemove, element)
+      }
+  }
+
+  function patch (parent, element, oldNode, newNode) {
     if (oldNode == null) {
-      element = parent.appendChild(createElementFrom(node))
+      var n = createElementFrom(newNode)
+      parent.insertBefore(n, element)
+      element = n
+
+    } else if (
+      newNode.tag !== oldNode.tag ||
+      typeof newNode !== typeof oldNode ||
+      typeof newNode === "string" && newNode !== oldNode
+    ) {
+      var n = createElementFrom(newNode)
+      parent.replaceChild(n, element)
+      element = n
+
+    } else if (oldNode.tag){
+      updateElementData(element, newNode.data, oldNode.data)
+      patchChildren(element, oldNode, newNode)
+    }
 
-    } else if (node == null) {
-      batch.push(function () {
-        if (oldNode.data && oldNode.data.onRemove) {
-          oldNode.data.onRemove(element)
-        }
-        parent.removeChild(element)
-      })
-    } else if (node.tag && node.tag === oldNode.tag) {
-      updateElementData(element, node.data, oldNode.data)
+    return element
+  }
+
+  function patchChildren(element, oldNode, newNode) {
+    // We need to index all keyed nodes in oldChildren in order to
+    // distinguish new keyed nodes from nodes to reuse in newChildren
+    var oldChildren = []
+    var oldKeys = {}
+    for (var i = 0; i < oldNode.children.length; i++) {
+      oldNode.children[i].data &&
+      oldNode.children[i].data.key &&
+      (oldKeys[oldNode.children[i].data.key] = true);
+      oldChildren[i] = [element.childNodes[i], oldNode.children[i]]
+    }
 
-      var len = node.children.length
-      var oldLen = oldNode.children.length
+    var isKeyed = function (node) {
+      return node.data && node.data.key && oldKeys[node.data.key]
```

Comment:
Actually I think we can have numbers. The number 0 as well. Not sure exactly how various browsers handle this, but in v8 at least (just popped open my node console and tested this)

    > var a = {}
    undefined
    > a[0] = 'foo'
    'foo'
    > a 
    {'0': 'foo'}
    > a[0]
    'foo'

Sure, it will *cast* numbers to strings for the representation. But as long as you are not using a mix of numbers and stringed numbers to key your children (and *that* is a really unlikely scenario), it should work without conflict